src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -0,0 +1,559 @@
+
+        offsets = new FrequencyRandomOffsets(transitLayer);
+    }
+
+    /** For each iteration, return the travel time to each transit stop */
+    public int[][] route () {
+        startClockTime = System.nanoTime();
+
+        if (saveAllStates) statesEachIteration = new ArrayList<>();
+
+        prefilterPatterns();
+
+        // compute number of minutes for scheduled search
+        int nMinutes = (request.toTime - request.fromTime) / DEPARTURE_STEP_SEC;
+
+        // how many monte carlo draws per minute of scheduled search to get desired total iterations?
+        int monteCarloDrawsPerMinute = (int) Math.ceil((double) request.monteCarloDraws / nMinutes);
+
+        LOG.info("Performing {} scheduled iterations each with {} Monte Carlo draws for a total of {} iterations",
+                nMinutes, monteCarloDrawsPerMinute, nMinutes * monteCarloDrawsPerMinute);
+
+        int[][] results = new int[transit.hasFrequencies ? nMinutes * monteCarloDrawsPerMinute : nMinutes][];
+        int currentIteration = 0;
+
+        // main loop over departure times
+        for (int departureTime = request.toTime - DEPARTURE_STEP_SEC, minute = nMinutes;
+             departureTime >= request.fromTime; departureTime -= DEPARTURE_STEP_SEC, minute--) {
+            if (minute % 15 == 0) LOG.info("  minute {}", minute);
+
+            updateDepartureTime(departureTime);
+
+            // run the search
+            int[][] resultsForMinute = runRaptorForMinute(monteCarloDrawsPerMinute);
+
+            // effectively final nonsense
+            final int finalDepartureTime = departureTime;
+            for (int[] resultsForIteration : resultsForMinute) {
+                // NB this copies the array, so we don't have issues with it being updated later
+                results[currentIteration++] = IntStream.of(resultsForIteration)
+                        .map(r -> r != RaptorWorker.UNREACHED ? r - finalDepartureTime : r)
+                        .toArray();
+            }
+        }
+
+        LOG.info("Search completed in {}s", (System.nanoTime() - startClockTime) / 1e9d);
+        LOG.info("Scheduled/bounds search: {}s", timeInScheduledSearch / 1e9d);
+        LOG.info("  - Scheduled search: {}s", timeInScheduledSearchTransit / 1e9d);
+        LOG.info("  - Frequency upper bounds: {}s", timeInScheduledSearchFrequencyBounds / 1e9d);
+        LOG.info("  - Transfers: {}s", timeInScheduledSearchTransfers / 1e9d);
+        LOG.info("Frequency search: {}s", timeInFrequencySearch / 1e9d);
+        LOG.info("  - Frequency component: {}s", timeInFrequencySearchFrequency / 1e9d);
+        LOG.info("  - Resulting updates to scheduled component: {}s", timeInFrequencySearchScheduled / 1e9d);
+        LOG.info("  - Transfers: {}s", timeInFrequencySearchTransfers / 1e9d);
+
+        return results;
+    }
+
+    /** Prefilter the patterns to only ones that are running */
+    private void prefilterPatterns () {
```

Comment:
My thought was that filtering might get the patterns into the CPU cache, but since they're references and they won't all be on the same page perhaps not.